File: android/cNews/src/main/java/com/androidpi/news/entity/News.java
package com.androidpi.news.entity;

import android.arch.persistence.room.Embedded;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Index;
import android.arch.persistence.room.PrimaryKey;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.google.gson.annotations.SerializedName;

/**
 * Created by jastrelax on 2018/8/10.
 */
@Entity(tableName = "news",
        indices = @Index(value = "url", unique = true))
public class News implements Parcelable{

    /**
     * source : {"id":null,"name":"Chinanews.com"}
     * author : chinanews
     * title : 卫健委通报陕西疫苗事件：疫苗未过期系登记错误
     * description : 调查还发现3个接种点存在接种门诊硬件设施不达标、接种工作人员数量不足、预防接种管理信息化程度低、监督检查不到位等问题。
     * url : http://www.chinanews.com/gn/2018/08-09/8595157.shtml
     * urlToImage : null
     * publishedAt : 2018-08-09T13:45:19Z
     */

    @PrimaryKey(autoGenerate = true)
    @NonNull
    private Long id;
    @Embedded(prefix = "source_")
    @SerializedName("source")
    private Source source;
    @SerializedName("author")
    private String author;
    @SerializedName("title")
    private String title;
    @SerializedName("description")
    private String description;
    @SerializedName("url")
    private String url;
    @SerializedName("urlToImage")
    private String urlToImage;
    @SerializedName("publishedAt")
    private String publishedAt;

    @NonNull
    public Long getId() {
        return id;
    }

    public void setId(@NonNull Long id) {
        this.id = id;
    }

    public Source getSource() {
        return source;
    }

    public void setSource(Source source) {
        this.source = source;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public String getUrlToImage() {
        return urlToImage;
    }

    public void setUrlToImage(String urlToImage) {
        this.urlToImage = urlToImage;
    }

    public String getPublishedAt() {
        return publishedAt;
    }

    public void setPublishedAt(String publishedAt) {
        this.publishedAt = publishedAt;
    }

    public static class Source implements Parcelable{
        /**
         * id : null
         * name : Chinanews.com
         */

        @SerializedName("id")
        private String id;
        @SerializedName("name")
        private String name;

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }


        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(this.id);
            dest.writeString(this.name);
        }

        public Source() {
        }

        protected Source(Parcel in) {
            this.id = in.readString();
            this.name = in.readString();
        }

        public static final Creator<Source> CREATOR = new Creator<Source>() {
            @Override
            public Source createFromParcel(Parcel source) {
                return new Source(source);
            }

            @Override
            public Source[] newArray(int size) {
                return new Source[size];
            }
        };
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(this.id);
        dest.writeParcelable(this.source, flags);
        dest.writeString(this.author);
        dest.writeString(this.title);
        dest.writeString(this.description);
        dest.writeString(this.url);
        dest.writeString(this.urlToImage);
        dest.writeString(this.publishedAt);
    }

    public News() {
    }

    protected News(Parcel in) {
        this.id = (Long) in.readValue(Long.class.getClassLoader());
        this.source = in.readParcelable(Source.class.getClassLoader());
        this.author = in.readString();
        this.title = in.readString();
        this.description = in.readString();
        this.url = in.readString();
        this.urlToImage = in.readString();
        this.publishedAt = in.readString();
    }

    public static final Creator<News> CREATOR = new Creator<News>() {
        @Override
        public News createFromParcel(Parcel source) {
            return new News(source);
        }

        @Override
        public News[] newArray(int size) {
            return new News[size];
        }
    };
}

File: news/src/main/java/cn/androidpi/news/entity/News.kt
package cn.androidpi.news.entity

import android.arch.persistence.room.ColumnInfo
import android.arch.persistence.room.Entity
import android.arch.persistence.room.Index
import android.arch.persistence.room.PrimaryKey
import android.os.Parcel
import android.os.Parcelable
import java.util.*

/**
 * 一条新闻。
 * Created by jastrelax on 2017/11/2.
 */

@Entity(tableName = "news",
        indices = arrayOf(Index(value = "news_id", unique = true)))
class News() : Parcelable {
    @PrimaryKey
    var id: Long? = null

    /**
     * 服务端的id
     */
    @ColumnInfo(name = "news_id")
    var newsId: String? = null

    /**
     * 分类
     */
    var category: String? = null

    /**
     * 原始标题
     */
    @ColumnInfo(name = "origin_title")
    var originTitle: String? = null

    /**
     * 发表时间
     */
    @ColumnInfo(name = "publish_time")
    var publishTime: String? = null

    /**
     * 来源名称
     */
    @ColumnInfo(name = "source_name")
    var sourceName: String? = null

    /**
     * 来源链接
     */
    @ColumnInfo(name = "source_url")
    var sourceUrl: String? = null

    /**
     * 标题
     */
    var title: String? = null

    /**
     * 链接
     */
    var url: String? = null

    /**
     * 关键词
     */
    var keywords: Array<String>? = null

    /**
     * 新闻图片链接
     */
    var images: Array<String>? = null

    /**
     * 上下文，关于新闻所在页面，板块等信息
     */
    var context: String? = null

    /**
     * 门户站点
     */
    var portal: String? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as News

        if (id != other.id) return false
        if (newsId != other.newsId) return false
        if (category != other.category) return false
        if (originTitle != other.originTitle) return false
        if (publishTime != other.publishTime) return false
        if (sourceName != other.sourceName) return false
        if (sourceUrl != other.sourceUrl) return false
        if (title != other.title) return false
        if (url != other.url) return false
        if (!Arrays.equals(keywords, other.keywords)) return false
        if (!Arrays.equals(images, other.images)) return false
        if (context != other.context) return false
        if (portal != other.portal) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + (newsId?.hashCode() ?: 0)
        result = 31 * result + (category?.hashCode() ?: 0)
        result = 31 * result + (originTitle?.hashCode() ?: 0)
        result = 31 * result + (publishTime?.hashCode() ?: 0)
        result = 31 * result + (sourceName?.hashCode() ?: 0)
        result = 31 * result + (sourceUrl?.hashCode() ?: 0)
        result = 31 * result + (title?.hashCode() ?: 0)
        result = 31 * result + (url?.hashCode() ?: 0)
        result = 31 * result + (keywords?.let { Arrays.hashCode(it) } ?: 0)
        result = 31 * result + (images?.let { Arrays.hashCode(it) } ?: 0)
        result = 31 * result + (context?.hashCode() ?: 0)
        result = 31 * result + (portal?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "News(id=$id, newsId=$newsId, category=$category, originTitle=$originTitle, publishTime=$publishTime, sourceName=$sourceName, sourceUrl=$sourceUrl, title=$title, url=$url, keywords=${Arrays.toString(keywords)}, images=${Arrays.toString(images)})"
    }

    constructor(source: Parcel) : this() {
        id = source.readLong()
        newsId = source.readString()
        category = source.readString()
        originTitle = source.readString()
        publishTime = source.readString()
        sourceName = source.readString()
        sourceUrl = source.readString()
        title = source.readString()
        url = source.readString()
        keywords = source.createStringArray()
        images = source.createStringArray()
        context = source.readString()
        portal = source.readString()
    }

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        dest.writeLong(id ?: -1)
        dest.writeString(newsId)
        dest.writeString(category)
        dest.writeString(originTitle)
        dest.writeString(publishTime)
        dest.writeString(sourceName)
        dest.writeString(sourceUrl)
        dest.writeString(title)
        dest.writeString(url)
        dest.writeStringArray(keywords)
        dest.writeStringArray(images)
        dest.writeString(context)
        dest.writeString(portal)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<News> = object : Parcelable.Creator<News> {
            override fun createFromParcel(source: Parcel): News = News(source)
            override fun newArray(size: Int): Array<News?> = arrayOfNulls(size)
        }
    }
}
